### ISSUE TITLE
Feature: Add static get method to Api-Gateway V1 REST Api

### ISSUE THREAD

**Issue by dror-weiss** [sst/sst#3978]
I would like to have one API gateway instanciated in one stage and references by another.

**Comment by dror-weiss** [comment]
@fwang , can you please review this again, or, if you think it's undoable, offer a workaround to accomplish the same thing in another way?

**Comment by jayair** [comment]
Sorry what did you want reviewed?

**Comment by dror-weiss** [comment]
@jayair , thanks.
It's basically adding `get` to an existing `ApiGatewayV1` that I can call from another deployment.

[This](https://github.com/sst/ion/compare/dev...dror-weiss:ion:feat/api-gateway-v1-ref#diff-6ecdaffb8923e896af590b41097cbb206b2577cc72acb16248fd247cf49f7054R1418) was the source branch of a PR that was closed due to the new SST repo. Sorry for the noisy changes.

**Comment by fwang** [comment]
@dror-weiss In ur snippet above, multiple SST stages share the same REST API, and each SST stage has their own REST API stage. I can imagine a (made up) use case where multiple SST stages share the same REST API stage. It's not clear the best way to support various use cases.

> offer a workaround to accomplish the same thing in another way

For a workaround, try create the API in one stage, and import in other stages.
ie. in one stage
```ts
const api  = new sst.aws.ApiGatewayV1("MyApi", {
  ...
});
api.route();
```

And in another stage
```ts
const api  = new sst.aws.ApiGatewayV1("MyApi", {
  ...,
  transform: {
    api: (args, opts) => {
      opts.import = "xxxx_API_ID_xxxx";
      opts.retainOnDelete = true;
    }
  }
});
api.route();
```

In general, we don't recommend [importing](https://sst.dev/docs/import-resources) the same resource to multiple stages. Here we set `retainOnDelete` to ensure the API can only get removed by the stage that created the API.

I haven't tried this.. but lemme know if it works for you.

**Comment by maxkoryukov** [comment]
@fwang it does not work.

and worse - `sst remove` will delete the existing resource (happened to me, and it was pretty painful)

here is the error message:

```
ApiGatewayRestApi sst:aws:ApiGatewayV1 → ApiGatewayRestApiResourceBabtrc aws:apigateway:Resource
sdk-v2/provider2.go:515: sdk.helper_schema: creating API Gateway Resource: operation error API Gateway: CreateResource, https response error StatusCode: 409, RequestID: 5869cd40-5724-45d1-9760-4625c80cc4db, ConflictException: Another resource with the same parent already has this name: api: provider=aws@6.72.0
```

versions:
```
sst@3.9.39
aws@6.72.0
```

**Comment by maxkoryukov** [comment]
> In general, we don't recommend [importing](https://sst.dev/docs/import-resources) the same resource to multiple stages. 

`sst` is a tool that helps to deploy the infrastructure. I have a huge number of resources created outside of SST, and for me there is no way to move them into SST stack because.. mostly because: why would a business pay me for writing the same architecture if the existing one works. So I need to "reference" most of the existing resources.

another philosophical question. here is the situation: I created an ApiGateway using terraform/serverless/cdk/sst/manually/pulumi/nameit/... - now I can go to the AWS UI and I **can** add a route manually in the UI. I can even add 5 routes to the existing ApiGatewayV1. My thought train was like: ok, i did it manually in 10 mins. Then SST (with constructs and the ability to inspect the existing stack) can help me to do it automatically every time i have changes in my code.

**but SST v3 doesn't have any tips/hints about how to add a `route` to the _existing_ ApiGateway**

Before I used to create 1 ApiGateway for the whole project (for all stages, and with "edge" - for all regions), and I used it for ALL deployments because it was easier to have one ApiGatewya with one root URL, and since ApiGateway forcibly adds `stage` to the URI-path - it was the way to to differentiate deployments.

So for me it is a big blocker that I can not `ApiGatewayV1.get()`. I would say: this "reference existing resources" for extending them (by adding triggers/routes/authorizers/other resources) is a big blocker when one has an existing stack but wants to add more resources using SST.

I like how fast one can deploy a Next.js app with SST. But I didn't expect it to be that difficult to discover that referencing an existing ApiGateway doesn't work. Or I don't know how to google/chatgpt/kapaai it

**Comment by jayair** [comment]
> So I need to "reference" most of the existing resources.

You misread what Frank was saying.

> In general, we don't recommend importing the same resource to multiple stages.

Firstly, he's talking about importing, not referencing. These are different concepts.

<img width="779" alt="Image" src="https://github.com/user-attachments/assets/b9d9843b-5544-47e0-b6c8-685c815afd43" />

Secondly, in the case you were trying to import. You don't want multiple stages being responsible for the same resource.

All that said, API Gateway and AWS in general can have some default guidelines that end up being bad design patters in the first place. And this is an example of that. You ideally should create multiple API Gateway projects for separate stages.

**Comment by maxkoryukov** [comment]
you are right about importing. and even more right about managing **the same resource** from different stages (or using different tools) - it leads to a fiasco pretty quickly.

but the ticket is about **referencing**, and that's what I needed. Just like in OP's situation, I also have an existing ApiGateway V1 (or "Rest" for some reason) and I need to add several new routes (preferably, using SST). It doesn't matter how - by creating a Route object with an extra param `apiId` OR by calling `api.route()` method after _referencing_ `api` - all are good if they work.

I didn't plan it to take days, but after about ~1 month trying to solve this, I still don't know if it is feasible. So, some time ago I moved from SST to another tool. I'm still sure SST is a great tool (we deploy sites in minutes), but now I need to hammer a nail, and SST seems to be a screwdriver.

**Comment by seanjb** [comment]
> REST API needs to create a `Deployment` resource after all routes are added. It needs to be aware of ALL the routes, and compute a unique hash based on routes' properties. When multiple SST apps are adding routes to the same API, each app will create a `Deployment` resource with partial information about the routes.

@fwang I would be comfortable with a simplified logic that removes the hashing logic and triggers a deploy only if any local route changes are made. 

That way the last service to add routes will be responsible for triggering a deploy. I believe Serverless framework followed this pattern. (in Serverless we would create API Gateway in one project - then import it as a resource in each other project). Each importing project would trigger a redeploy whenever it added or changed config on the API (routes)

**Comment by seanjb** [comment]
Another specific use case for a centralised shared API is authorised caching. If I have multiple API gateways then my authoriser cache is not shared amongst them. 

This means each first request to an endpoint managed of a different gateway will need reauthorized adding latency and cost.
